Add auto TLS enricher

jimmidyson opened the pull request:
Fixes #697 

Advice needed on how to best apply this so it only applies to OpenShift manifests.

Also, how to add this as an optional enricher that can be added to `<enricher><includes><include>...` in a project? Easy to add as a default enricher, but this should be opt-in, rather than a default.

jimmidyson commented:
And yes I need to add docs.

jimmidyson commented:
So this has grown a bit because I had to refactor the init container stuff, had to update `VolumePermissionEnricher` so added a test for that too.

This needs to wait to merge until @rhuss has sorted out the default/none profiles as discussed on IRC.

rhuss commented:
@jimmidyson hope that I have something running this evening, but we need to test this thoroughly since the algorithm changed quite a bit. Let you know when I'm done.

rhuss commented:
I have something running , but unfortunately not yet for prime time. Still missing tests and docs (and the definition of the 'default' profile). Work is happening on #709

cmoulliard commented:
What this TLE enricher is supposed to do (generate a self signed certificate + keystore) ? Is there a document describing the use case implemented ?

jimmidyson commented:
See the referenced issue #697 plus the OpenShift docs at https://docs.openshift.org/latest/dev_guide/secrets.html#service-serving-certificate-secrets

rhuss commented on enricher/api/src/main/java/io/fabric8/maven/enricher/api/BaseEnricher.java:
```diff
@@ -114,4 +127,32 @@ protected boolean isOpenShiftMode() {
+    }
+
+    protected void addInitContainer(ObjectMetaBuilder metadata, JSONObject initContainer) {
+        String initContainerAnnotation = metadata.getAnnotations().get(INIT_CONTAINER_ANNOTATION);
```
annotations are always non-null, right ? 

I always get confused with when using builder and when using object directly. I.e. I would have expected here a 'editAnnotations()' or so. But I remember darkly that for lists one always use the 'direct' objects.

jimmidyson replied:
This is retrieving current annotation, hence `getAnnotations`, so that the two can be merged. Should always be non-null.

rhuss commented on enricher/api/src/main/java/io/fabric8/maven/enricher/api/BaseEnricher.java:
```diff
@@ -114,4 +127,32 @@ protected boolean isOpenShiftMode() {
+        metadata.addToAnnotations(INIT_CONTAINER_ANNOTATION, initContainers.toString());
+    }
+
+    protected void addInitContainer(PodTemplateSpecBuilder builder, JSONObject initContainer) {
```
What the difference between both 'addInitContainer()' methods ?

jimmidyson replied:
Removed the unused one :)

rhuss commented on enricher/api/src/main/java/io/fabric8/maven/enricher/api/BaseEnricher.java:
```diff
@@ -114,4 +127,32 @@ protected boolean isOpenShiftMode() {
+                : new JSONArray(initContainerAnnotation);
+        initContainers.put(initContainer);
+
+        builder.editMetadata().addToAnnotations(INIT_CONTAINER_ANNOTATION, initContainers.toString()).endMetadata();
```
Ah, here it is :) That's what I expected how to deal with annotations on builders.

jimmidyson replied:
:)

rhuss commented on enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/AutoTLSEnricher.java:
```diff
@@ -0,0 +1,190 @@
+    public AutoTLSEnricher(EnricherContext buildContext) {
+        super(buildContext, ENRICHER_NAME);
+
+        this.secretName = getConfig(Config.tlsSecretName);
```
you could use directly

```java
this.secretName = getConfig(Config.tlsSecretName, getProject().getArtifactId() + "-tls");
```

jimmidyson replied:
Done.

rhuss commented on enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/AutoTLSEnricher.java:
```diff
@@ -0,0 +1,190 @@
+    @Override
+    public Map<String, String> getAnnotations(Kind kind) {
+        if (!isOpenShiftMode()) {
+            return super.getAnnotations(kind);
```
You should probably return `null` directly here since your know that your enricher doesn't want to add any annotations in this case.

Actually I would collapse this method in sth like

```java
if (!isOpenShiftMode() || !kind.isService()) {
   return null;
}
return Collections.singletonMap(AUTOTLS_ANNOTATION_KEY, this.secretName);
```

jimmidyson replied:
Done.

rhuss commented on enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/AutoTLSEnricher.java:
```diff
@@ -0,0 +1,190 @@
+    @Override
+    public void addMissingResources(KubernetesListBuilder builder) {
+        if (!isOpenShiftMode()) {
+            super.addMissingResources(builder);
```
wouldn't call super here as it is not necessary. 

Enrichers are not chained, they are called individually and are hence independent. The `BaseEnricher` was introduced merely to allow an implementer to ommit implementing certain interface methods if she is not interested in them. Actually its more like the former Swing Adapters used for event handling.

jimmidyson replied:
Done.

rhuss commented on enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/AutoTLSEnricher.java:
```diff
@@ -0,0 +1,190 @@
+    @Override
+    public void adapt(KubernetesListBuilder builder) {
+        if (!isOpenShiftMode()) {
+            super.adapt(builder);
```
dito

jimmidyson replied:
Done.

rhuss commented on enricher/standard/src/main/test/java/io/fabric8/maven/enricher/standard/VolumePermissionEnricherTest.java:
```diff
@@ -0,0 +1,126 @@
+
+            VolumePermissionEnricher enricher = new VolumePermissionEnricher(context);
+
+            PodTemplateBuilder ptb = new PodTemplateBuilder().withNewMetadata().and()
```
just out of curiosity: what does 'and()' do ?

jimmidyson replied:
Just like `end...` but more fluent (reads better & means you don't have to remember what kind you're editing).

jimmidyson commented:
This is now working thanks to @rhuss's brilliant profile work.

rhuss commented on enricher/api/src/main/java/io/fabric8/maven/enricher/api/BaseEnricher.java:
```diff
@@ -114,4 +127,32 @@ protected boolean isOpenShiftMode() {
+        }
+    }
+
+    protected void addInitContainer(ObjectMetaBuilder metadata, JSONObject initContainer) {
```
Seems like that this method is never used ?

jimmidyson replied:
Removed.

jimmidyson commented:
Do you want me to squash?

BTW the changes to VolumePermissionEnricher were to reuse the init container editing between two enrichers that contribute init containers. I think as a side effect it may actually have fixed #772  btw - checking now.

rhuss commented:
Yeah, would make sense to squash, as I often have to juggle with the v3.1.80-redhat branch, too. hence would prefer one commit per feature.

jimmidyson commented:
Hmm can't recreate #772 anyway... odd.

jimmidyson commented:
Squashed commits.

rhuss commented:
thanks ;-) Looks good. Now only the documentation is left (a simple table with options is enough) and the entry in `enricher-default`, then we should merge.

jimmidyson commented:
Sorry - I've added some docs for the options now.

jimmidyson commented:
@rawlingsj I refactore the VolumePermissionEnricher as part of this PR to share handling of init containers in BaseEnricher. This also included a change to the visitor approach for updates to resources. I've just rebased including your recent PR to fix VolumePermissionEnricher - would you mind casting your eye over it to see I haven't broken anything?

I've also added a test for it so I can add the case you've just fixed to that test if you'd like - can you just explain what the problem was so I can update the test case appropriately?

jimmidyson commented:
@rawlingsj Don't worry - figured out the test stuff. Works a treat :)

rawlingsj commented:
@jimmidyson ah cool - yeah LGTM, the main bit for me was only adding the initcontainer annotation for PVC volumes which you have, so all good!

rhuss commented:
Thanks ! LGTM, too

[merge]

rhuss commented on enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/AutoTLSEnricher.java:
```diff
@@ -0,0 +1,180 @@
+
+    @Override
+    public void addMissingResources(KubernetesListBuilder builder) {
+        if (!isOpenShiftMode()) {
```
While trying to fix other stuff, stumbled over this base `isOpenShiftMode()` method which has been introduced lately.

@jimmidyson @jstrachan @rawlingsj : Just a heads up (and also it was used for adding docker env vars to pod envs when using s2i binary build): It won't work as you might expect, since we always create kubernetes.yml and openshift.yml, regardless of the platform mode (auto detected or with -Dfabric8.mode). When the platform mode is used in an enricher it will create different kubernetes.yml and openshift.yml depending on the mode, but it still creates both files. And you can perfectly apply an openshift.yml to OpenShift which has been created with a `mvn -Dfabric8.mode=kubernetes fabric8:resource` (or when its running in auto detect mode and not being connected to an active OpenShift context). 

This is also dangerous when you refer to these files included in dependencies as it now matters with which mode these dependencies has been build.

The solution is of course again described in #678 so that actually kubernetes and openshift objects are not converted into each other in an after thought but by calling each enricher twice, once for kubernetes and once for openshift, independent on the fabric8.mode in use (which however still is needed for deciding how to do the build).

Until then this might work or not and for sake of simplicity I even will take this call over to fix the VolumePermissionEnricher to not add init containers for OpenShift.
